### PR TITLE
Add thread safe allocator for Fortran

### PR DESCRIPTION
**Pull request by davidbeckingsale**

### CODE REVIEW

File: examples/cookbook/recipe_thread_safe.F [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+      pool = rm%make_allocator_pool("POOL",         &
+                                    base_allocator, &
+                                    512_8*1024_8,   &
+                                    1024_8)
+
+      thread_safe_allocator = rm%make_allocator_thread_safe("THREAD_SAFE_POOL", &
+                                                            pool)
+
+      call thread_safe_allocator%allocate(array, [10])
+end program umpire_f_pinned_pool
```
Comment (mcfadden8) [inline examples/cookbook/recipe_thread_safe.F]: Kudos for the recipe for this!!

File: tests/integration/strategy_tests.cpp [hunk 2]
```diff
@@ -618,8 +654,66 @@ TEST(ThreadSafeAllocator, Host)
...
+      "thread_safe_allocator", rm.getAllocator("DEVICE"));
+
+  constexpr int N = 16;
+  std::vector<void*> thread_allocs{N};
+  std::vector<std::thread> threads;
+
+  for (std::size_t i = 0; i < N; ++i)
+  {
+    threads.push_back(
+        std::thread([=, &allocator, &thread_allocs] { thread_allocs[i] = allocator.allocate(1024); }));
```
Comment (mcfadden8) [inline tests/integration/strategy_tests.cpp]: I would suggest something a little more stressful.  I think the loop below may induce more things happening in parallel within the `allocator` object:

```suggestion
        std::thread([=, &allocator, &thread_allocs] {
            for ( j = 0; j < N; ++j)
            {
                thread_allocs[i] = allocator.allocate(1024);
                ASSERT_NE(thread_allocs[i]);
                allocator.deallocate(thread_allocs[i]);
                ASSERT_THROW({ allocator.deallocate(thread_allocs[i]); });
                thread_allocs[i] = allocator.allocate(1024);
        }));
```

Thoughts?

Comment (davidbeckingsale) [reply tests/integration/strategy_tests.cpp]: I agree!

File: tests/integration/strategy_tests.cpp [hunk 3]
```diff
@@ -618,8 +654,74 @@ TEST(ThreadSafeAllocator, Host)
...
+
+  constexpr int N = 16;
+  std::vector<void*> thread_allocs{N};
+  std::vector<std::thread> threads;
+
+  for (std::size_t i = 0; i < N; ++i)
+  {
+    threads.push_back(
+        std::thread([=, &allocator, &thread_allocs] {
+            for ( j = 0; j < N; ++j)
```
Comment (mcfadden8) [inline tests/integration/strategy_tests.cpp]: Sorry!
```suggestion
            for ( int j = 0; j < N; ++j)
```

**Pull request merged**